src/rpc.rs

```diff
@@ -10,82 +10,131 @@
-        let zenoh_key = if let Some(sink) = request.attributes.sink.0.clone() {
-            self.to_zenoh_key_string(&source, Some(&sink))
-        } else {
-            self.to_zenoh_key_string(&source, None)
+impl RpcClient for ZenohRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
+        call_options: CallOptions,
+        payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        // Get data and format from UPayload
+        let mut payload_data = Bytes::new();
+        let mut payload_format = UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED;
+        if let Some(payload) = payload {
+            payload_format = payload.payload_format();
+            payload_data = payload.payload();
+        }
+
+        // Get source UUri
+        let source_uri = self.uri_provider.get_source_uri();
+
+        let attributes = UAttributes {
+            type_: UMessageType::UMESSAGE_TYPE_REQUEST.into(),
+            id: Some(call_options.message_id().unwrap_or_else(UUID::build)).into(),
+            priority: call_options
+                .priority()
+                .unwrap_or(UPriority::UPRIORITY_UNSPECIFIED)
+                .into(),
+            source: Some(source_uri.clone()).into(),
+            sink: Some(method.clone()).into(),
+            ttl: Some(call_options.ttl()),
+            token: call_options.token(),
+            payload_format: payload_format.into(),
+            ..Default::default()
         };
 
-        // Create UAttributes and put into Zenoh user attachment
-        let attributes = *request.attributes.0.clone().ok_or_else(|| {
-            let msg = "Invalid UAttributes".to_string();
-            error!("{msg}");
-            UMessageError::AttributesValidationError(UAttributesError::ParsingError(msg))
-        })?;
-        let Ok(attachment) = UPClientZenoh::uattributes_to_attachment(&attributes) else {
+        // Get Zenoh key
+        let zenoh_key = self
+            .transport
+            .to_zenoh_key_string(&source_uri, Some(&method));
+
+        // Put UAttributes into Zenoh user attachment
+        let Ok(attachment) = UPTransportZenoh::uattributes_to_attachment(&attributes) else {
             let msg = "Unable to transform UAttributes to user attachment in Zenoh".to_string();
             error!("{msg}");
-            return Err(UMessageError::AttributesValidationError(
-                UAttributesError::ParsingError(msg),
-            ));
-        };
-
-        // Get the data from UPayload
-        let value = if let Some(payload) = request.payload {
-            Value::new(payload.to_vec().into())
-        } else {
-            Value::new(vec![].into())
+            return Err(ServiceInvocationError::Internal(msg));
         };
 
         // Send the query
         let mut getbuilder = self
+            .transport
             .session
             .get(&zenoh_key)
-            .with_value(value)
+            .with_value(payload_data.as_ref())
```

Comment:
Is there a penalty in invoking `with_value` with empty payload?

Reply:
TBH, I don't know. But I agree that we can avoid invoking `with_value` while there is no payload.